From 226b897f8d3515da27dfabb945833e8b7ce637e5 Mon Sep 17 00:00:00 2001
From: Tobias Schramm <t.schramm@t-sys.eu>
Date: Wed, 10 Feb 2021 15:32:27 +0100
Subject: [PATCH 2/6] sunxi: Add support for OtterCast Audio V2

The OtterCast Audio V2 is an Allwinner S3 based SBC designed by
Jan Henrik Hemsing. It is primarily designed as a audio streaming
device, featuring a high quality I2S DAC, as well as a single fast
ethernet port, 2.4 GHz b/g/n WIFi and Bluetooth 4.2.

Signed-off-by: Tobias Schramm <t.schramm@t-sys.eu>
---
 arch/arm/dts/Makefile                        |   1 +
 arch/arm/dts/sun8i-s3-ottercast-audio-v2.dts | 148 +++++++++++++++++++
 configs/ottercast_audio_v2_defconfig         |  17 +++
 3 files changed, 166 insertions(+)
 create mode 100644 arch/arm/dts/sun8i-s3-ottercast-audio-v2.dts
 create mode 100644 configs/ottercast_audio_v2_defconfig

diff --git a/arch/arm/dts/Makefile b/arch/arm/dts/Makefile
index af2842a0f8..9aa29d2463 100644
--- a/arch/arm/dts/Makefile
+++ b/arch/arm/dts/Makefile
@@ -598,6 +598,7 @@ dtb-$(CONFIG_MACH_SUN8I_R40) += \
 	sun8i-r40-bananapi-m2-ultra.dtb \
 	sun8i-v40-bananapi-m2-berry.dtb
 dtb-$(CONFIG_MACH_SUN8I_V3S) += \
+	sun8i-s3-ottercast-audio-v2.dtb \
 	sun8i-s3-pinecube.dtb \
 	sun8i-v3s-licheepi-zero.dtb
 dtb-$(CONFIG_MACH_SUN50I_H5) += \
diff --git a/arch/arm/dts/sun8i-s3-ottercast-audio-v2.dts b/arch/arm/dts/sun8i-s3-ottercast-audio-v2.dts
new file mode 100644
index 0000000000..5442fa8208
--- /dev/null
+++ b/arch/arm/dts/sun8i-s3-ottercast-audio-v2.dts
@@ -0,0 +1,148 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR X11)
+/*
+ * Copyright 2021 Tobias Schramm <t.schramm@t-sys.eu>
+ */
+
+/dts-v1/;
+#include "sun8i-v3.dtsi"
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	model = "OtterCast Audio V2";
+	compatible = "jhh,ottercast-audio-v2", "allwinner,sun8i-s3";
+
+	aliases {
+		serial0 = &uart1;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led1 {
+			label = "led:usr";
+			gpios = <&pio 1 4 GPIO_ACTIVE_HIGH>; /* PB4 */
+		};
+
+		led2 {
+			label = "led:pwr";
+			gpios = <&pio 1 5 GPIO_ACTIVE_HIGH>; /* PB5 */
+			default-state = "on";
+		};
+	};
+
+	wifi_pwrseq: wifi_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&pio 4 19 GPIO_ACTIVE_LOW>; /* PE19 WL-PMU-EN */
+		post-power-on-delay-ms = <200>;
+	};
+};
+
+&emac {
+	phy-handle = <&int_mii_phy>;
+	phy-mode = "mii";
+	status = "okay";
+};
+
+&i2c0 {
+	status = "okay";
+
+	axp209: pmic@34 {
+		compatible = "x-powers,axp203",
+			     "x-powers,axp209";
+		reg = <0x34>;
+		interrupt-controller;
+		#interrupt-cells = <1>;
+	};
+};
+
+&i2c1 {
+	status = "okay";
+};
+
+&lradc {
+	vref-supply = <&reg_ldo2>;
+	status = "okay";
+/*
+	button-200 {
+		label = "Setup";
+		linux,code = <KEY_SETUP>;
+		channel = <0>;
+		voltage = <190000>;
+	};
+*/
+};
+
+&mmc0 {
+	vmmc-supply = <&reg_dcdc3>;
+	bus-width = <4>;
+	cd-gpios = <&pio 5 6 GPIO_ACTIVE_LOW>;
+	status = "okay";
+};
+
+&mmc1 {
+	vmmc-supply = <&reg_dcdc3>;
+	vqmmc-supply = <&reg_dcdc3>;
+	mmc-pwrseq = <&wifi_pwrseq>;
+	bus-width = <4>;
+	non-removable;
+	status = "okay";
+};
+
+&pio {
+	vcc-pd-supply = <&reg_dcdc3>;
+	vcc-pe-supply = <&reg_ldo3>;
+};
+
+#include "axp209.dtsi"
+
+&ac_power_supply {
+	status = "okay";
+};
+
+&reg_dcdc2 {
+	regulator-always-on;
+	regulator-min-microvolt = <1100000>;
+	regulator-max-microvolt = <1100000>;
+	regulator-name = "vdd-sys-cpu-ephy";
+};
+
+&reg_dcdc3 {
+	regulator-always-on;
+	regulator-min-microvolt = <3300000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-name = "vcc-3v3";
+};
+
+&reg_ldo1 {
+	regulator-name = "vdd-rtc";
+};
+
+&reg_ldo2 {
+	regulator-always-on;
+	regulator-min-microvolt = <3000000>;
+	regulator-max-microvolt = <3000000>;
+	regulator-name = "avcc";
+};
+
+&uart1 {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart1_pg_pins>;
+};
+
+&uart2 {
+	status = "okay";
+};
+
+&pio {
+	uart1_pg_pins: uart1-pg-pins {
+		pins = "PG6", "PG7";
+		function = "uart1";
+	};
+};
diff --git a/configs/ottercast_audio_v2_defconfig b/configs/ottercast_audio_v2_defconfig
new file mode 100644
index 0000000000..9283e6ed0c
--- /dev/null
+++ b/configs/ottercast_audio_v2_defconfig
@@ -0,0 +1,17 @@
+CONFIG_ARM=y
+CONFIG_ARCH_SUNXI=y
+CONFIG_SPL=y
+CONFIG_MACH_SUN8I_V3S=y
+CONFIG_SUNXI_DRAM_DDR3_1333=y
+CONFIG_DRAM_CLK=504
+CONFIG_DRAM_ODT_EN=y
+CONFIG_I2C0_ENABLE=y
+CONFIG_DEFAULT_DEVICE_TREE="sun8i-s3-ottercast-audio-v2"
+CONFIG_SPL_I2C_SUPPORT=y
+# CONFIG_NETDEVICES is not set
+CONFIG_AXP209_POWER=y
+CONFIG_AXP_DCDC2_VOLT=1100
+CONFIG_AXP_DCDC3_VOLT=3300
+CONFIG_AXP_ALDO2_VOLT=3000
+CONFIG_AXP_ALDO3_VOLT=0
+CONFIG_CONS_INDEX=2
-- 
2.30.1

